### PR TITLE
FINERACT-1806: Advanced Charge-off Expense Accounting - "Advanced Accounting Rule" takes priority (if configured)

### PR DESCRIPTION
**Pull request by kulminsky**
Advanced Charge-off Expense Accounting - "Advanced Accounting Rule" takes priority (if configured)

## Description

Advanced Charge-off Expense Accounting - "Advanced Accounting Rule" takes priority (if configured)

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Review by galovics — CHANGES_REQUESTED** [review]
Let's add tests to cover the "priority".

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java [hunk 1]
```diff
@@ -229,14 +230,27 @@ private void createJournalEntriesForChargeOff(LoanDTO loanDTO, LoanTransactionDT
...
-        if (principalAmount != null && principalAmount.compareTo(BigDecimal.ZERO) > 0) {
-            if (isMarkedFraud) {
-                populateCreditDebitMaps(loanProductId, principalAmount, paymentTypeId, AccrualAccountsForLoan.LOAN_PORTFOLIO.getValue(),
-                        AccrualAccountsForLoan.CHARGE_OFF_FRAUD_EXPENSE.getValue(), glAccountBalanceHolder);
-            } else {
-                populateCreditDebitMaps(loanProductId, principalAmount, paymentTypeId, AccrualAccountsForLoan.LOAN_PORTFOLIO.getValue(),
-                        AccrualAccountsForLoan.CHARGE_OFF_EXPENSE.getValue(), glAccountBalanceHolder);
+
+        // need to fetch if there are account mappings (always one)
+        Integer chargeOffReasonCode = loanDTO.getChargeOffReasonCode();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java]: This is not a code, it is a code value

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java [hunk 2]
```diff
@@ -229,14 +230,27 @@ private void createJournalEntriesForChargeOff(LoanDTO loanDTO, LoanTransactionDT
...
+
+        ProductToGLAccountMapping mapping = helper.getChargeOffMappingByCode(chargeOffReasonCode);
+        if (mapping != null) {
+            GLAccount accountCredit = this.helper.getLinkedGLAccountForLoanProduct(loanProductId,
+                    AccrualAccountsForLoan.LOAN_PORTFOLIO.getValue(), paymentTypeId);
+            glAccountBalanceHolder.addToCredit(accountCredit, principalAmount);
+            // Resolve Debit
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java]: No need for this comment

Comment (kulminsky) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java (hunk 1 above)
Comment (kulminsky) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java]: Done

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**